libs/application-generic/src/usecases/normalize-variables/normalize-variables.usecase.ts

```diff
@@ -0,0 +1,128 @@
+import { Injectable } from '@nestjs/common';
+import {
+  SubscriberEntity,
+  SubscriberRepository,
+  TenantEntity,
+  TenantRepository,
+} from '@novu/dal';
+import { IMessageFilter } from '@novu/shared';
+import { IFilterVariables } from '../../utils/filter-processing-details';
+import { CachedEntity } from '../../services/cache/interceptors/cached-entity.interceptor';
+import { buildSubscriberKey } from '../../services/cache/key-builders/entities';
+import { ConditionsFilterCommand } from '../conditions-filter';
+
+/**
+ * This service class is responsible for normalizing the variables used within the message filtering process.
+ * Normalization in this context refers to ensuring all necessary data is present for filter evaluation.
+ *
+ * It achieves this by:
+ *  1. Checking if subscriber and tenant information are provided in the command itself.
+ *  2. If missing, it tries to infer them from the filters and job data (if available).
+ *  3. Finally, it fetches the complete subscriber and tenant entities from the database if necessary.
+ *
+ * By providing a normalized set of variables, this service simplifies filter evaluation and promotes code clarity.
+ */
+@Injectable()
+export class NormalizeVariables {
+  constructor(
+    private subscriberRepository: SubscriberRepository,
+    private tenantRepository: TenantRepository
+  ) {}
+
+  public async execute(command: ConditionsFilterCommand) {
+    const filterVariables: IFilterVariables = {};
+
+    const combinedFilters = [
+      command.step,
+      ...(command.step?.variants || []),
+    ].flatMap((variant) => (variant?.filters ? variant?.filters : []));
+
+    filterVariables.subscriber = await this.fetchSubscriberIfMissing(
+      command,
+      combinedFilters
+    );
+    filterVariables.tenant = await this.fetchTenantIfMissing(
+      command,
+      combinedFilters
+    );
+    filterVariables.payload = command.variables?.payload
+      ? command.variables?.payload
+      : command.job?.payload ?? undefined;
+
+    filterVariables.step = command.variables?.step ?? undefined;
+    filterVariables.actor = command.variables?.actor ?? undefined;
+
+    return filterVariables;
+  }
+  private async fetchSubscriberIfMissing(
+    command: ConditionsFilterCommand,
+    filters: IMessageFilter[]
+  ): Promise<SubscriberEntity | undefined> {
+    if (command.variables?.subscriber) {
+      return command.variables.subscriber;
+    }
+
+    const subscriberFilterExist = filters?.find((filter) => {
+      return filter?.children?.find((item) => item?.on === 'subscriber');
+    });
+
+    if (subscriberFilterExist && command.job) {
+      return (
+        (await this.getSubscriberBySubscriberId({
+          subscriberId: command.job.subscriberId,
+          _environmentId: command.environmentId,
+        })) ?? undefined
+      );
+    }
+
+    return undefined;
+  }
+
+  private async fetchTenantIfMissing(
```

Comment:
<!-- Nitpicks are trivial preference-based requests. These should be non-blocking by nature.  -->
🤓 **nitpick (non-blocking):** ‏ Is the `**IfMissing()` pattern of naming precise enough? It kind of implies that if it is not missing it won't be fetched. Maybe just keep the `fetch<THING>()` as the name (or exchange `fetch` for `get` or `find`)

Reply:
hmm well if `command.variables?.tenant` exists we will return it and won't execute any retrievals.

Reply:
true, but this exact element of the process is missing in the function name, the name only concerns itself with the situation when `command.variables?.tenant` is missing.

Not much value in dying on this hill though, in general I think the name is acceptable